Add spawnProcessDetached

FreeSlave opened the pull request:
spawnProcessDetached is function similar to spawnProcess, but user does not need to wait on spawned process to avoid zombies (resource leak). Instead of returning Pid instance spawnProcessDetached returns integer pid via pointer. Rationale: user should not be able to wait on detached process.

Currently only Posix version is implemented. It uses double fork technique and setsid to detach the new process from its caller. Since this implementation reuses a lot of stuff from spawnProcess I moved some implementation details to separate private functions to share them between spawnProcess and spawnProcessDetached.

THIS IS WIP
I'll add more unittests and Windows version later. For now say if you like the idea.

CyberShadow commented:
I believe that on Windows, you simply have to close the process handle as soon as the process is created to effectively "detach" the process. (This is not the same as a detached console window.)

Can this be implemented as a `Config` flag instead? I'm not sure about adding an entire overload. Then it can be used for `spawnShell` as well. `Pid` will likely need a few changes to make waiting on a detached processes an explicit error.

FreeSlave commented:
> I believe that on Windows, you simply have to close the process handle as soon as the process is created to effectively "detach" the process.

I believe so. This is another reason why I did not want to use Pid class. Process handle is part of Pid class on Windows. Closing it will make osHandle be able to be null. So users will need to check for osHandle not being null if they want to use it. That differs from the current situation when osHandle is always not null. And it's also inconsistent with Posix, where osHandle is same as pid and therefore always valid. Do you think it is ok?

CyberShadow commented:
> So users will need to check for osHandle not being null if they want to use it.

I think it's more of not doing a thing that doesn't make sense to do.

> And it's also inconsistent with Posix, where osHandle is same as pid and therefore always valid.

If you spawn a detached process, technically there is no guarantee that the PID remains valid for any amount of time - the process may exit on its own and the OS may assign the PID to another process. So, if we are to reuse `Pid`, probably the best course is a private `bool owned` field (which is false for detached processes), and `kill` and `wait` require that `owned` is true.

CC @schveiguy @kyllingstad

FreeSlave commented:
> kill and wait require that owned is true

Why would kill require owned is true? You can kill any process via task manager or kill program which are clearly does not own this process.

CyberShadow commented:
Because it might not be the process you actually intend to kill - that process may have exited a long time ago.

FreeSlave commented:
Got it.
I'm still not sure that Pid class can be re-used here. There was that PR https://github.com/dlang/phobos/pull/5086 that introduced possibility to have non-owned Pid instances. And it was reverted later. Probably because it complicates the usage of other functions (like already said `wait` and `kill`).

CyberShadow commented:
#5086 had more problems than that, though (killing processes being an intrinsically platform-dependent operation; needing to call `OpenProcess` on Windows). I think overloading `Pid` here makes more sense than adding an overload set.

kyllingstad commented:
> I think overloading `Pid` here makes more sense than adding an overload set.

I very much agree with this.

FreeSlave commented:
Added Pid.owned propert and Config.detached constant.
Also added separate unittest to check non-ability of waiting/killing not owned process. It runs OK, but I have this message at terminal
```
/bin/sh: 0: Can't open /tmp/std.process temporary file 76b5fac2-57f8-4f5b-88dc-9fd27c32ae73
```

CyberShadow commented on std/process.d:
```diff
@@ -338,7 +339,131 @@ version(Posix) private enum InternalError : ubyte
+version(Posix)
+{
+private:
+    const(char)[] getExecutablePath(const(char)[] name)
```
Since there is no longer an overload set, these refactorings aren't needed any more, right? If so, can they be undone?

CyberShadow commented on std/process.d:
```diff
@@ -1316,6 +1428,12 @@ enum Config
+
+    /**
+    Spawn process in detached state. This removes the need in calling
+    $(LREF wait) to clean up the process resources.
```
Should have a note here that calling `wait` or `kill` with the resulting `Pid` is invalid.

CyberShadow commented on std/process.d:
```diff
@@ -1358,6 +1476,15 @@ final class Pid
+    This is $(D true) if process was spawned without $(LREF Config.detached), $(D false)
+    otherwise. You can't $(LREF wait) for or $(LREF kill) process you don't own.
+    */
+    @property bool owned() const @safe pure nothrow
```
No need to make this public. I can't think of many cases that programs would mix owned and non-owned Pids and require distinguishing them. Making at an implementation detail also allows changing the underlying implementation in the future, if necessary.

CyberShadow commented on std/process.d:
```diff
@@ -1374,6 +1501,8 @@ private:
     int performWait(bool block) @trusted
     {
+        import std.exception : enforceEx;
+        enforceEx!ProcessException(_owned, "Attempt to wait for the not owned process");
```
Consider "Can't wait on a detached process"

CyberShadow commented on std/process.d:
```diff
@@ -1422,6 +1551,8 @@ private:
         int performWait(bool block) @trusted
         {
+            import std.exception : enforceEx;
+            enforceEx!ProcessException(_owned, "Attempt to wait for the not owned process");
```
Ditto

CyberShadow commented on std/process.d:
```diff
@@ -1503,7 +1647,7 @@ Signal codes are defined in the $(D core.sys.posix.signal) module
 
 Throws:
-$(LREF ProcessException) on failure.
+$(LREF ProcessException) on failure or on attempt to wait for not owned process.
```
Consider "detached" or "non-child" instead of "not owned", the "owned" term should be an implementation detail

CyberShadow commented on std/process.d:
```diff
@@ -1558,7 +1702,7 @@ Returns:
 
 Throws:
-$(LREF ProcessException) on failure.
+$(LREF ProcessException) on failure or on attempt to wait for not owned process.
```
ditto

CyberShadow commented on std/process.d:
```diff
@@ -1636,6 +1780,7 @@ assert(wait(pid) == -SIGKILL); // Negative return value on POSIX!
 
 Throws:
 $(LREF ProcessException) on error (e.g. if codeOrSignal is invalid).
+    or on attempt to kill not owned process.
```
ditto

CyberShadow commented on std/process.d:
```diff
@@ -1652,6 +1797,8 @@ void kill(Pid pid)
 void kill(Pid pid, int codeOrSignal)
 {
+    import std.exception : enforceEx;
+    enforceEx!ProcessException(pid.owned(), "Attempt to kill not owned process");
```
ditto

CyberShadow commented:
> It runs OK, but I get this message at terminal

Is this an expected and unavoidable effect of the unit test, or do you need time/assistance in tracking it down?

FreeSlave commented:
> Is this an expected and unavoidable effect of the unit test, or do you need time/assistance in tracking it down?

It should not happen. It's simple TestScript usage. But I can't understand why it happens.

FreeSlave commented:
Ok, I think I got it. TestScript destructor runs (effectively removing the file) before file execution. But kernel somehow manages to find out that file should have ran with /bin/sh. Weird.

CyberShadow commented on std/process.d:
```diff
@@ -416,15 +418,21 @@ private Pid spawnProcessImpl(in char[][] args,
+        else
+            throw ProcessException.newFromErrno("Could not create pipe to get process pid");
+    }
+    scope(exit) if (config & Config.detached) close(pidPipe[0]);
```
Any reason to not reuse the same pipe used for communicating the result (`forkPipe`)?

FreeSlave replied:
After the second fork started the first and the second forks run in parallel. I can't ensure that pid will be written to the pipe before possible errors.

CyberShadow replied:
I see. Please add this explanation as a comment somewhere.

CyberShadow replied:
I guess calling `getpid` from the second child is also a possibility, but I guess it's not worth the change.

FreeSlave replied:
There's also a possibility that the second fork fails. In this case we would need to pass fictitious pid from the first fork to adjust.

andralex replied:
make these considerations a nice comment

FreeSlave replied:
Already considered in the comment before `int[2] pidPipe;`

CyberShadow commented on std/process.d:
```diff
@@ -450,94 +465,133 @@ private Pid spawnProcessImpl(in char[][] args,
+                close(workDirFD);
             }
-            close(workDirFD);
         }
```
Why change the working directory in the second fork? Wouldn't it be simpler to change it in the first fork regardless of `Config.detached`?

CyberShadow commented on std/process.d:
```diff
@@ -450,94 +465,133 @@ private Pid spawnProcessImpl(in char[][] args,
-                pfds[i].fd = i + 3;
-                pfds[i].events = 0;
-                pfds[i].revents = 0;
+                setsid();
```
Hmm, isn't `setsid` supposed to be called from the first fork (so that the session leader is killed)?

FreeSlave replied:
You're right. Will change.

FreeSlave replied:
I found that setsid makes some apps run wrong when starting from terminal without setting stdin to /dev/null.
E.g. run `setsid openarena` from terminal. It shows a window, but does render anything.
Should we do anything about that?

CyberShadow replied:
Unfortunately I am not sufficiently familiar with the subject to confidently answer that.

From looking online, I noticed that `setsid` is something that is often done when double-forking, but not prescriptively so. Perhaps it only makes sense for non-interactive daemon processes?

FreeSlave replied:
If we don't setsid, the spawned process will belong to the same session so pkill -s '$sid' will send signal to this process too.

CyberShadow replied:
When is that a problem?

FreeSlave replied:
Oh, well, not all apps belong to the same session actually. Only those that was ran from desktop. If it runs from terminal, session is different.
The point is that user may want to have it in another session or in the same. I think it's better to leave to user to decide.

CyberShadow replied:
OK, so correct me if I'm wrong, but as I understand the most commonly-observed difference will be that ^C will not kill the spawned process if it's in the same session.

In which case, it does sound like it's worthy of a `Flag` addition, with the Windows implementation being mapped to the `CREATE_NEW_PROCESS_GROUP` process creation flag.

CyberShadow replied:
It may be worth leaving it for a separate PR.

FreeSlave replied:
^C is sent only to the foreground processes of session. And session has distinction between foreground and background processes only if it has connection to controlling terminal.
Even if doing setsid via flag it's still not clear if user want the new process to be a session leader (i.e. call setsid in the second fork) or not (call setsid in the first fork and let it die).

> It may be worth leaving it for a separate PR.
So shall I remove setsid for now?

CyberShadow replied:
Thanks for the clarification. I think I'm out of my depth here. Perhaps someone more knowledgeable than me about POSIX process sessions can chime in, @kyllingstad @schveiguy @jmdavis ?

FreeSlave replied:
I've done some experiments. Login via ssh (just plain login without tmux, screen or similar) and spawned some script process in detached state without calling to setsid. Logout and re-login again, the process still ran. So I guess setsid is not necessary after all.

CyberShadow commented on std/process.d:
```diff
@@ -1359,6 +1459,11 @@ final class Pid
     }
 
 private:
+    @property bool owned() const @safe pure nothrow
```
No point in a private property, just name the field `owned`.

CyberShadow commented on std/process.d:
```diff
@@ -1709,6 +1834,25 @@ void kill(Pid pid, int codeOrSignal)
+    This leads to the annoying message like "/bin/sh: 0: Can't open /tmp/std.process temporary file" to appear when running tests.
+    It does not happen in unittests with non-detached processes because we always wait() for them to finish.
+    */
+    Thread.sleep(dur!"msecs"(5));
```
I think this is fine, but I'd give it a more generous timeout of 100 msecs or so just to exclude race conditions.

FreeSlave replied:
I used 5 msecs, because there's the same sleep in the other unittest under version(Android), having a comment about Android killing sleeping applications.

CyberShadow replied:
That's fine, we don't have CI for Android; I'm more concerned about Windows, where process creation is very slow. If the Android maintainer finds that this test is causing them trouble then they can fix it as needed. (Also, I was pretty sure Android kills idle apps, not processes...)

CyberShadow commented:
Could you please rebase on top of the latest master? That should fix CircleCI.

FreeSlave commented:
Rebased. Is any other change needed?

CyberShadow commented:
Ah, a changelog entry or linked bug as per @dlang-bot's comment. I think that's all if there is no further feedback.

FreeSlave commented:
I've added changelog https://github.com/FreeSlave/phobos/commit/fa8a73f47b953049ee8751bffb31401d33d79344, but github does not see the changes. Probably because of the previous rebase.

CyberShadow commented:
@wilzbach Any idea why changelog is not showing up in DAutoTest diff here?

wilzbach commented:
> @wilzbach Any idea why changelog is not showing up in DAutoTest diff here?

Yes, it seems like Martin has added this file and thus it doesn't get generated:

https://github.com/dlang/dlang.org/blob/master/changelog/2.075.0_pre.dd

Should we bump the version at dlang.org? Probably a good idea to do this for all version files after the master -> stable branch-off?

CyberShadow commented:
> Should we bump the version at dlang.org? Probably a good idea to do this for all version files after the master -> stable branch-off?

Sounds like a good idea.

CyberShadow commented on changelog/std-process-config-detached.dd:
```diff
@@ -0,0 +1,5 @@
+
+Config.detached allows to $(REF_ALTTEXT spawn process, spawnProcess, std, process) independently from the caller.
+No need to wait on the process spawned that way - no zombie will left after process exits.
+Attempts to $(REF wait, std, process) on or $(REF kill, std, process) detached processes will throw.
```
Some copyediting:

`Config.detached` allows $(REF_ALTTEXT spawning processes, spawnProcess, std, process) which run independently from the current process. There is no need to `wait` on the processes created with this flag, and no $(HTTPS en.wikipedia.org/wiki/Zombie_process, zombie process) will be left after they exit.
Attempts to call $(REF wait, std, process) or $(REF kill, std, process) on detached processes will throw.

JackStouffer commented:
Does this really need Andrei's approval? It doesn't really add a new symbol.

wilzbach commented:
> Does this really need Andrei's approval? It doesn't really add a new symbol.

Well there's `Config.detached`, but yeah I'm a bit frustrated about the huge bottleneck that this is causing as well. Maybe
- we should let the bot send @andralex a mail every week about open PRs which require his approval
- move away from this single point of failure and redefine this policy to "at least two codeowners need to approve a new symbol" (or require the @andralex label only for modules he cares about, i.e. for which he is a code owner)

andralex commented on std/process.d:
```diff
@@ -339,7 +340,8 @@ version(Posix) private enum InternalError : ubyte
     chdir,
-    getrlimit
+    getrlimit,
+    doubleFork
```
add a comma so the next addition blames properly

andralex commented on std/process.d:
```diff
@@ -415,17 +417,30 @@ private Pid spawnProcessImpl(in char[][] args,
+    // but we don't have a direct access to the second fork pid from the caller side thus use a pipe.
+    // We also can't reuse forkPipe for that purpose
+    // because we can't predict the order in which pid and possible error will be written
+    // since the first and the second forks will run in parallel.
```
looks like a multiline comment

andralex commented on std/process.d:
```diff
@@ -415,17 +417,30 @@ private Pid spawnProcessImpl(in char[][] args,
+        if (core.sys.posix.unistd.pipe(pidPipe) == 0)
+            setCLOEXEC(pidPipe[1], true);
+        else
+            throw ProcessException.newFromErrno("Could not create pipe to get process pid");
```
Simplify flow:

```d
if (core.sys.posix.unistd.pipe(pidPipe) != 0)
    throw ...;
setCLOEXEC(pidPipe[1], true);
```

andralex commented on std/process.d:
```diff
@@ -465,80 +490,110 @@ private Pid spawnProcessImpl(in char[][] args,
+            // In the case that stderr is redirected to stdout, we need
+            // to backup the file descriptor since stdout may be redirected
+            // as well.
+            if (stderrFD == STDOUT_FILENO)  stderrFD = dup(stderrFD);
```
double space?

FreeSlave replied:
It was like this in previous implementation

andralex commented on std/process.d:
```diff
@@ -465,80 +490,110 @@ private Pid spawnProcessImpl(in char[][] args,
+                immutable maxToClose = maxDescriptors - 3;
+
+                // Call poll() to see which ones are actually open:
+                pollfd* pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
```
auto

FreeSlave replied:
Again, you're asking to change things that were like this before me. Well, I can, for sure.

andralex replied:
Yes and much appreciated!

andralex commented on std/process.d:
```diff
@@ -548,18 +603,29 @@ private Pid spawnProcessImpl(in char[][] args,
         auto status = InternalError.noerror;
         auto readExecResult = core.sys.posix.unistd.read(forkPipe[0], &status, status.sizeof);
+        // Save error number just in case if subsequent "waitpid" fails and overrides errno
+        auto lastError = .errno;
```
immutable

andralex commented on std/process.d:
```diff
@@ -571,15 +637,28 @@ private Pid spawnProcessImpl(in char[][] args,
-
             if (readExecResult == error.sizeof)
                 throw ProcessException.newFromErrno(error, errorMsg);
             else
```
delete

andralex commented on std/process.d:
```diff
@@ -571,15 +637,28 @@ private Pid spawnProcessImpl(in char[][] args,
+            if (read(pidPipe[0], &actualPid, id.sizeof) == id.sizeof)
+                id = actualPid;
+            else
+                throw ProcessException.newFromErrno("Could not read from pipe to get detached process id");
```
simplify flow

andralex commented on std/process.d:
```diff
@@ -689,8 +768,15 @@ private Pid spawnProcessImpl(in char[] commandLine,
+        CloseHandle(pi.hProcess);
+        return new Pid(pi.dwProcessId);
+    }
+    else
```
felete

schveiguy replied:
delete? ;)

FreeSlave replied:
Delete what? Can you elaborate?

andralex replied:
I mean delete the `else` line.

andralex commented:
>move away from this single point of failure and redefine this policy to "at least two codeowners need to approve a new symbol" (or require the @andralex label only for modules he cares about, i.e. for which he is a code owner)

Please let me be the point of failure for the sake of better checks and balances of library consistency. In several prior cases it has happened that design sensibilities somewhat foreign to D/Phobos have entered the standard library, to everybody's displeasure. Thanks.

schveiguy commented on std/process.d:
```diff
@@ -1009,40 +1095,58 @@ version (Posix) @system unittest
+            wait(pid);
+        else
+            // We need to wait a little to ensure that the process has finished and data was written to files
+            Thread.sleep(dur!"msecs"(200));
```
Wait longer. There's no need to vie for performance here. Wait 2 seconds instead of .2

schveiguy replied:
Also, stylistic change, you can put `200.msecs` instead of `dur!"msecs"(200)`

FreeSlave replied:
I would prefer unittests run as fast as they could. But ok.

schveiguy commented on std/process.d:
```diff
@@ -1712,6 +1849,25 @@ void kill(Pid pid, int codeOrSignal)
+    This leads to the annoying message like "/bin/sh: 0: Can't open /tmp/std.process temporary file" to appear when running tests.
+    It does not happen in unittests with non-detached processes because we always wait() for them to finish.
+    */
+    Thread.sleep(dur!"msecs"(100));
```
Longer time here too.

schveiguy commented:
ping @CyberShadow any idea why the DAutoTest fails?

schveiguy commented:
> Bugzilla references

@wilzbach this looks invalid to me, this has nothing to do with RedBlackTree. I don't actually see a bugzilla reference in any commit, so not sure why this showed up.

CyberShadow commented:
> ping @CyberShadow any idea why the DAutoTest fails?

Please see https://github.com/dlang/phobos/pull/5639#issuecomment-316838940 / https://github.com/dlang/phobos/pull/5626#issuecomment-316285183 / https://github.com/dlang/phobos/pull/5625#issuecomment-315921337 / https://github.com/dlang/phobos/pull/5619#issuecomment-315758105 / https://github.com/dlang/phobos/pull/5617#issuecomment-315668792.

TL;DR: GitHub bug.

schveiguy commented:
@CyberShadow thanks, here's a followup: while this bug exists, what is the correct response, seeing as that test is required for merging? Should I just ping you/@wilzbach, or is there a way I can fix it myself?

wilzbach commented:
> while this bug exists, what is the correct response, seeing as that test is required for merging? Should I just ping you/@wilzbach, or is there a way I can fix it myself?

AFAICT the only safe way is to create a new PR.
It also needs to be a different commit sha :/

We can overwrite it manually if the auto-tester passes, but there is always the slight risk of breaking Jenkins or CircleCI though. However, at least for CircleCi we get a warning mail afterwards, so I think we can start to risk it. Creating a new PR all the time, is really annoying.

wilzbach commented:
> @wilzbach this looks invalid to me, this has nothing to do with RedBlackTree. I don't actually see a bugzilla reference in any commit, so not sure why this showed up.

I would guess that it's due to the GitHub API sending rubbish.

wilzbach commented:
> We can overwrite it manually if the auto-tester passes, but there is always the slight risk of breaking Jenkins or CircleCI though. However, at least for CircleCi we get a warning mail afterwards, so I think we can start to risk it. Creating a new PR all the time, is really annoying.

Tried it this time - let's hope for the best. We really need to get GitHub to fix their bugs though..
